### PR TITLE
win,unix: uv_buf_init() accepts size_t

### PR DESCRIPTION
**Pull request by trevnorris**
Change the signature of `uv_buf_init()` from `(char*, unsigned int)` to
`(char*, size_t)`. This is to accommodate the fact that the `len` property
of `uv_buf_t` is a `size_t`, and fixes a problem where `len` is truncated
for large `base`.

Fixes: https://github.com/libuv/libuv/issues/1233

R=@bnoordhuis 
R=@saghul

### CODE REVIEW

**Comment by cjihrig** [comment]
Docs need to be updated as well (probably just the function signature, and maybe a `versionchanged::` note).

**Comment by saghul** [comment]
What @cjihrig said.

In addition: on 64bit Windows `size_t` is `unsigned long long`, but the `WSABUF` structure to which we cast our `uv_buf_t` declares the length field as `unsigned long`: https://msdn.microsoft.com/nl-nl/library/windows/desktop/ms741542(v=vs.85).aspx not sure if this can be a problem in practice.

I'd like the test to have an ifdef and size check after casting to `struct iovec` and `WSABUF` respectively.

**Comment by trevnorris** [comment]
@saghul Only allowing an `unsigned long` is probably a problem. I've been testing on my machine by reading/writing 6GB data files. Which would seem to be impossible to address using a `WSABUF`.

**Comment by trevnorris** [comment]
@saghul I think I've added the tests that you've requested. If there are any others, let me know.

**Comment by santigimeno** [comment]
This should be tagged as `v2` right?

**Comment by bnoordhuis** [comment]
Yes, done.

File: test/test-buf.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/* Copyright (c) 2017, Trevor Norris <trev.norris@gmail.com>
```
Comment (cjihrig) [inline test/test-buf.c]: Should this be libuv collaborators?

**Comment by cjihrig** [comment]
I fixed up this commit, but the test fails on zOS because `SIZE_MAX` is not defined. @jBarz could you take a look at https://github.com/cjihrig/libuv/commit/538ea4c416637abd1664db4d7314b5fe10c8bb7f?

**Comment by cjihrig** [comment]
Ping @jBarz regarding `SIZE_MAX` on zOS. I'd like to move forward with this PR.

**Comment by jBarz** [comment]
looking into it, few mins

**Comment by jBarz** [comment]
```diff
@@ -191,7 +192,7 @@
           ],
         }],
         [ 'OS == "zos"', {
-          'cflags': [ '-qxplink' ],
+          'cflags': [ '-qlanglvl=stdc99', '-qxplink' ],
           'ldflags': [ '-qxplink' ],
         }],
         ['uv_library=="shared_library"', {
@@ -255,7 +256,7 @@
           ]
         }],
         [ 'OS == "zos"', {
-          'cflags': [ '-qxplink' ],
+          'cflags': [ '-qlanglvl=stdc99', '-qxplink' ],
           'ldflags': [ '-qxplink' ],
         }],
         ['uv_library=="shared_library"', {
```

**Comment by cjihrig** [comment]
@jBarz thank you!

Closing this PR in favor of #1806.

**Pull request closed without merging**